controllers/containerizedworkload_controller.go

```diff
@@ -0,0 +1,156 @@
+*/
+
+package controllers
+
+import (
+	"context"
+	"github.com/pkg/errors"
+	"time"
+
+	cpv1alpha1 "github.com/crossplaneio/crossplane-runtime/apis/core/v1alpha1"
+	"github.com/go-logr/logr"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	oamv1alpha2 "github.com/oam-dev/core-resource-controller/api/v1alpha2"
+)
+
+const (
+	oamReconcileWait = 30 * time.Second
+)
+
+// Reconcile error strings.
+const (
+	errRenderWorkload  = "cannot render workload"
+	errUpdateStatus    = "cannot apply status"
+	errApplyDeployment = "cannot apply the deployment"
+	errApplyService    = "cannot apply the service"
+	errGCDeployment    = "cannot clean up stale deployments"
+	errScaleDeployment = "cannot scale the deployment"
+)
+
+// ContainerizedWorkloadReconciler reconciles a ContainerizedWorkload object
+type ContainerizedWorkloadReconciler struct {
+	client.Client
+	Log    logr.Logger
+	Scheme *runtime.Scheme
+}
+
+// +kubebuilder:rbac:groups=core.oam.dev,resources=containerizedworkloads,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=core.oam.dev,resources=containerizedworkloads/status,verbs=get;update;patch
+func (r *ContainerizedWorkloadReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	ctx := context.Background()
+	log := r.Log.WithValues("containerizedworkload", req.NamespacedName)
+	log.Info("Reconcile container workload")
+
+	var workload oamv1alpha2.ContainerizedWorkload
+	if err := r.Get(ctx, req.NamespacedName, &workload); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+	log.Info("Get the workload")
+
+	deploy, err := r.renderWorkload(ctx, &workload)
+	if err != nil {
+		workload.Status.SetConditions(cpv1alpha1.ReconcileError(errors.Wrap(err, errRenderWorkload)))
+		log.Error(err, "Failed to render a deployment")
+		return reconcile.Result{RequeueAfter: oamReconcileWait}, errors.Wrap(r.Status().Update(ctx, &workload),
+			errUpdateStatus)
+	}
+	log.Info("Successfully rendered a deployment", "deployment", deploy.Name)
+
+	// server side apply, only the fields we set are touched
+	applyOpts := []client.PatchOption{client.ForceOwnership, client.FieldOwner(workload.ObjectMeta.Name)}
+	if err := r.Patch(ctx, deploy, client.Apply, applyOpts...); err != nil {
+		workload.Status.SetConditions(cpv1alpha1.ReconcileError(errors.Wrap(err, errApplyDeployment)))
+		log.Error(err, "Failed to apply to a deployment")
+		return reconcile.Result{RequeueAfter: oamReconcileWait}, errors.Wrap(r.Status().Update(ctx, &workload),
+			errUpdateStatus)
+	}
+	log.Info("Successfully applied a deployment", "UID", deploy.UID)
+
+	// record the new deployment
+	if len(workload.Status.Resources) == 0 {
+		workload.Status.Resources = make(map[types.UID]oamv1alpha2.ResourceReference)
+	}
```

Comment:
https://github.com/kubernetes/community/blob/6112b/contributors/devel/sig-architecture/api-conventions.md#lists-of-named-subobjects-preferred-over-maps

Kubernetes API convention is to avoid maps of objects anywhere in the API.

Reply:
Keep it as a map makes the logic much straightforward (and that's another reason I redefined workload structure).  With that said, it's not a big deal, I can change it.

Reply:
In this case, can I redefine the workloadStatus to have a two-dimensional array instead? It's much cleaner to separate out the two resources and a list of list approach seems to be the blessed one?

Reply:
Could you give a quick example of the 2D array you're suggesting? I know it's suboptimal to implement, but I believe a list of (named) objects is the blessed pattern, rather than a list of lists.

Reply:
oh, it basically contains a slice of UIDs for each type of resource. ie. one array of UIDS for service type and one array of UIDs for deployment